File: airesources/CSharp/Halite2/hlt/Position.cs
using System;

namespace Halite2.hlt {

    public class Position {

        private double xPos;
        private double yPos;

        public Position(double xPos, double yPos) {
            this.xPos = xPos;
            this.yPos = yPos;
        }

        public double GetXPos() {
            return xPos;
        }

        public double GetYPos() {
            return yPos;
        }

        public double GetDistanceTo(Position target) {
            double dx = xPos - target.GetXPos();
            double dy = yPos - target.GetYPos();
            return Math.Sqrt(Math.Pow(dx, 2) + Math.Pow(dy, 2));
        }
        
        public virtual double GetRadius() {
            return 0;
        }

        public int OrientTowardsInDeg(Position target) {
            return Util.AngleRadToDegClipped(OrientTowardsInRad(target));
        }

        public double OrientTowardsInRad(Position target) {
            double dx = target.GetXPos() - xPos;
            double dy = target.GetYPos() - yPos;

            return Math.Atan2(dy, dx) + 2 * Math.PI;
        }

        public Position GetClosestPoint(Position target) {
            double radius = target.GetRadius() + Constants.MIN_DISTANCE_FOR_CLOSEST_POINT;
            double angleRad = target.OrientTowardsInRad(this);

            double x = target.GetXPos() + radius * Math.Cos(angleRad);
            double y = target.GetYPos() + radius * Math.Sin(angleRad);

            return new Position(x, y);
        }

        public override bool Equals(Object o) {
            if (this == o) 
                return true;            

            if (o == null || GetType() != o.GetType())
                return false;
            
            Position position = (Position)o;

            if (position == null)
                return false;

            return Equals(position.xPos, xPos) && Equals(position.yPos, yPos);
        }

        public override int GetHashCode() {
            return base.GetHashCode();
        }

        public override string ToString() {
            return "Position(" + xPos + ", " + yPos + ")";
        }
    }
}

File: airesources/Scala/hlt/Position.scala
package hlt

class Position(val xPos: Double, val yPos: Double) {
  def getDistanceTo(target: Position): Double = {
    val dx = xPos - target.getXPos
    val dy = yPos - target.getYPos
    Math.sqrt(Math.pow(dx, 2) + Math.pow(dy, 2))
  }

  def orientTowardsInDeg(target: Position): Int = Util.angleRadToDegClipped(orientTowardsInRad(target))

  def getClosestPoint(target: Entity): Position = {
    val MIN_DISTANCE = 3
    val radius = target.radius + MIN_DISTANCE
    val angleRad = target.orientTowardsInRad(this)
    val dx = target.getXPos + radius * Math.cos(angleRad)
    val dy = target.getYPos + radius * Math.sin(angleRad)
    new Position(dx, dy)
  }

  def getYPos: Double = yPos

  def orientTowardsInRad(target: Position): Double = {
    val dx = target.getXPos - xPos
    val dy = target.getYPos - yPos
    Math.atan2(dy, dx) + 2 * Math.PI
  }

  def getXPos: Double = xPos

  override def toString: String = "Position(" + xPos + ", " + yPos + ")"

  override def equals(other: Any): Boolean = other match {
    case that: Position =>
      (that canEqual this) &&
        xPos == that.xPos &&
        yPos == that.yPos
    case _ => false
  }

  def canEqual(other: Any): Boolean = other.isInstanceOf[Position]

  override def hashCode(): Int = {
    val state = Seq(xPos, yPos)
    state.map(_.hashCode()).foldLeft(0)((a, b) => 31 * a + b)
  }
}